internal/app/cloudwatcher/node_stats.go

```diff
@@ -0,0 +1,458 @@
+	elastic "github.com/olivere/elastic/v7" // Elasticsearch client.
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/cloudwatch"
+
+	"github.com/mintel/elasticsearch-asg/pkg/es" // Extensions to the Elasticsearch client.
+)
+
+var (
+	// errInconsistentNodes is returns by newNodeStats when
+	// the inputs refer to different Elasticsearch nodes.
+	errInconsistentNodes = errors.New("inconsistent nodes")
+)
+
+// JVMHeapStats represents stats about the JVM
+// heap of an Elasticsearch node.
+type JVMHeapStats struct {
+	MaxBytes  int64
+	UsedBytes int64
+}
+
+// NodeStats represents stats about an Elasticsearch node.
+type NodeStats struct {
+	Name  string
+	Roles []string
+
+	// Node is excluded from shard allocation.
+	ExcludedFromAllocation bool
+
+	// Count of vCPUs.
+	VCPUs int
+
+	// Linux load.
+	Load1m  float64
+	Load5m  float64
+	Load15m float64
+
+	// JVM heap stats.
+	JVMHeapStats JVMHeapStats
+	JVMHeapPools map[string]JVMHeapStats
+
+	// Elasticsearch disk size in bytes.
+	FilesystemTotalBytes int64
+
+	// Elasticsearch disk available bytes.
+	FilesystemAvailableBytes int64
+}
+
+// NewNodeStats returns a new nodeStats based on the
+// responses from various APIs. It returns
+// ErrInconsistentNodes if the responses are for different
+// nodes.
+func NewNodeStats(
+	s *elastic.NodesStatsNode,
+	i *EC2Instance,
+	transient *es.ShardAllocationExcludeSettings,
+	persistent *es.ShardAllocationExcludeSettings,
+) (*NodeStats, error) {
+	if s.Name != i.ID {
+		return nil, errInconsistentNodes
+	}
+
+	excluded := (transient.HasName(s.Name) ||
+		transient.HasIP(s.IP) ||
+		transient.HasHost(s.Host) ||
+		persistent.HasName(s.Name) ||
+		persistent.HasIP(s.IP) ||
+		persistent.HasHost(s.Host))
+	if !excluded {
+		for k, v := range s.Attributes {
+			if transient.HasAttr(k, fmt.Sprint(v)) || persistent.HasAttr(k, fmt.Sprint(v)) {
+				excluded = true
+				break
+			}
+		}
+	}
+
+	roles := append([]string(nil), s.Roles...)
+	sort.Strings(roles)
+
+	n := &NodeStats{
+		Name:                   s.Name,
+		Roles:                  roles,
+		ExcludedFromAllocation: excluded,
+		VCPUs:                  i.VCPUs,
+		JVMHeapStats: JVMHeapStats{
+			MaxBytes:  s.JVM.Mem.HeapMaxInBytes,
+			UsedBytes: s.JVM.Mem.HeapUsedInBytes,
+		},
+		JVMHeapPools:             make(map[string]JVMHeapStats, len(s.JVM.Mem.Pools)),
+		Load1m:                   s.OS.CPU.LoadAverage["1m"],
+		Load5m:                   s.OS.CPU.LoadAverage["5m"],
+		Load15m:                  s.OS.CPU.LoadAverage["15m"],
+		FilesystemTotalBytes:     s.FS.Total.TotalInBytes,
+		FilesystemAvailableBytes: s.FS.Total.AvailableInBytes,
+	}
+	for k, v := range s.JVM.Mem.Pools {
+		n.JVMHeapPools[k] = JVMHeapStats{
+			MaxBytes:  v.MaxInBytes,
+			UsedBytes: v.UsedInBytes,
+		}
+	}
+
+	return n, nil
+}
+
+// HasRole returns true if the node has a particular role.
+func (s *NodeStats) HasRole(role string) bool {
+	switch role {
+	case "all":
+		return true
+	case "coordinate":
```

Comment:
Why "coordinate"?  What does this mean?

Reply:
"Coordinate" comes from the [Elasticsearch docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-node.html#coordinating-node). Every Elasticsearch node has the implicit role of coordinating traffic i.e. routing requests to other nodes, so a "coordinate" node is one that has no other roles. I'll add a comment.

- [ ] Add a comment about coordinating nodes.